doc/source/templates/05_dreambooth_finetuning/dreambooth/train.py

```diff
@@ -48,7 +66,54 @@ def get_target(scheduler, noise, latents, timesteps):
     raise ValueError(f"Unknown prediction type {pred_type}")
 
 
-def load_models(config, cuda):
+def add_lora_layers(unet, text_encoder):
+    unet_lora_attn_procs = {}
+    unet_lora_parameters = []
+    for name, attn_processor in unet.attn_processors.items():
+        cross_attention_dim = (
+            None
+            if name.endswith("attn1.processor")
+            else unet.config.cross_attention_dim
+        )
+        if name.startswith("mid_block"):
+            hidden_size = unet.config.block_out_channels[-1]
+        elif name.startswith("up_blocks"):
+            block_id = int(name[len("up_blocks.")])
+            hidden_size = list(reversed(unet.config.block_out_channels))[block_id]
+        elif name.startswith("down_blocks"):
+            block_id = int(name[len("down_blocks.")])
+            hidden_size = unet.config.block_out_channels[block_id]
+
+        if isinstance(
+            attn_processor,
+            (AttnAddedKVProcessor, SlicedAttnAddedKVProcessor, AttnAddedKVProcessor2_0),
+        ):
+            lora_attn_processor_class = LoRAAttnAddedKVProcessor
+        else:
+            lora_attn_processor_class = (
+                LoRAAttnProcessor2_0
+                if hasattr(F, "scaled_dot_product_attention")
+                else LoRAAttnProcessor
+            )
+
+        module = lora_attn_processor_class(
+            hidden_size=hidden_size,
+            cross_attention_dim=cross_attention_dim,
+            rank=LORA_RANK,
+        )
+        unet_lora_attn_procs[name] = module
+        unet_lora_parameters.extend(module.parameters())
+
+    unet.set_attn_processor(unet_lora_attn_procs)
+
+    text_lora_parameters = LoraLoaderMixin._modify_text_encoder(
+        text_encoder, dtype=torch.float32, rank=LORA_RANK
+    )
+
+    return unet_lora_parameters, text_lora_parameters
```

Comment:
Naming confusion: Are these returns `unet_with_lora`, and `text_enc_with_lora`? Or just the parameters?

Reply:
These only *returns* parameters and the model is modified *in place*.